easybuild/easyblocks/c/code_saturne.py

```diff
@@ -0,0 +1,167 @@
+##
+# Copyright 2009-2018 Ghent University
+#
+# This file is part of EasyBuild,
+# originally created by the HPC team of Ghent University (http://ugent.be/hpc/en),
+# with support of Ghent University (http://ugent.be/hpc),
+# the Flemish Supercomputer Centre (VSC) (https://www.vscentrum.be),
+# Flemish Research Foundation (FWO) (http://www.fwo.be/en)
+# and the Department of Economy, Science and Innovation (EWI) (http://www.ewi-vlaanderen.be/en).
+#
+# https://github.com/easybuilders/easybuild
+#
+# EasyBuild is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation v2.
+#
+# EasyBuild is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with EasyBuild. If not, see <http://www.gnu.org/licenses/>.
+##
+"""
+EasyBuild support for building and installing Code_Saturne,
+implemented as an easyblock.
+
+@author: Metin Cakircali (Juelich Supercomputing Centre, FZJ GmbH)
+"""
+
+import os
+
+from easybuild.framework.easyblock import EasyBlock
+from easybuild.framework.easyconfig.default import CUSTOM
+from easybuild.tools.filetools import apply_regex_substitutions
+from easybuild.tools.run import run_cmd
+from easybuild.tools.systemtools import get_shared_lib_ext
+
+import easybuild.tools.environment as env
+
+
+class EB_Code_underscore_Saturne(EasyBlock):
+    """Support for building and installing Code_Saturne."""
+
+    @staticmethod
+    def extra_options(extra_vars=None):
+        """Extra easyconfig parameters specific to Code_Saturne."""
+        extra_vars = EasyBlock.extra_options(extra_vars)
+
+        # add some custom easyconfig parameters specific to Code_Saturne
+        extra_vars.update({
+            'debug': [False, "Build the debug version.", CUSTOM],
+            'slurm': [False, "Build for the slurm workload manager.", CUSTOM],
+        })
+
+        return extra_vars
+
+    def configure_step(self):
+        """Configure step for Code_Saturne."""
+
+        self.log.info("Configuration step is running...")
+
+        # only use the opt flags
+        env.setvar("CFLAGS", os.environ['OPTFLAGS'])
+        env.setvar("CXXFLAGS", os.environ['OPTFLAGS'])
+        env.setvar("FCFLAGS", os.environ['OPTFLAGS'])
+
+        self.log.info("Running ./sbin/bootstrap ...")
+        cmd = './sbin/bootstrap'
+        run_cmd(cmd, log_all=True, simple=True, log_output=True)
+
+        cmd = ' '.join([
+            self.cfg['preconfigopts'],
+            './configure',
+            '--prefix=' + self.installdir,
+            '--without-modules',
```

Comment:
What does `--without-modules` imply exactly?

Can we make this a bit more flexible, by letting people use `with_modules = True` in the easyconfig for example (with the default being `False`)?

Reply:
Code_Saturne reads the "loaded" modules at the time of build and loads them later when you run the software. This behavior is same as EB. I don't see why collide 2 tools to manage module loads at the same time. This should stay as is.

Reply:
What you say is true and makes sense if you think about compilers and MPIs in a hierarchy, but in a flat module naming scheme things are different. Loading the `Code_Saturne` module would load the dependencies, but one can swap the compiler and the MPI runtime without unloading `Code_Saturne`. Similarly, even in a hierarchy, the same can happen with other dependencies. If you have 2 versions of a dependency, you can load `Code_Saturne` and then swap the dependency, and potentially run into issues. Maybe it is not a bad idea to enable the use of module detection here.

Reply:
What kind of issue?
how do you hide EB module and developers modules from Code_Saturne during build?

Reply:
You can't hide those modules in an easy way, but not everyone sets up the installation environment using modules like we do.

> What kind of issue?

The kind of issue that `Code_Saturne` tries to avoid by saving the list of modules and using them at run time I suppose. Incompatible libraries, options missing, etc.

Reply:
> You can't hide those modules in an easy way, but not everyone sets up the installation environment using modules like we do.

So, no solution for us?

> 
> > What kind of issue?
> 
> The kind of issue that `Code_Saturne` tries to avoid by saving the list of modules and using them at run time I suppose. Incompatible libraries, options missing, etc.

It comes to the same point; --with-modules is for setups without EB. IMO, If things mess up by swapping dependencies, then it should be addressed in EB not in software or any workaround in block files.

Reply:
> So, no solution for us?

Sure. As I see things, the solution for us is to use `--without-modules`. That would work fine in our setup.

> It comes to the same point; --with-modules is for setups without EB

No, it's not. `--with-modules` is for setups that:

1. Do not depend on modules to set up the general software installation environment
2. Have a flat module naming scheme

You can have both things in EB setups

Reply:
@damianam maybe you don't know what Code_Saturne does. Again, "--without-modules works" for every setup. If you have flat naming setup, then you have a problem but I suggest a useful feature that EB does not allow one to swap dependencies on such setups. Removing this option and setting it in config is just unnecessary and cumbersome approach and you can't inline with mainstream or push such configs to mainstream. We already have such a diverse setup, better minimize it.

Reply:
As far as I understand the discussion, `--without-modules` makes sense? EB handles this, Code_Saturne should not interfere?

Reply:
Please don't confuse EB with the actual module system in use.
EB just builds software and creates modules in one of several module naming systems.
What the actual module system does is not EB's problem.
Lmod for instance can be setup to refuse swapping altogether or allow it.